### PR TITLE
KAFKA-14543: Move LogOffsetMetadata to storage module

### PR DESCRIPTION
**Pull request by mimaison**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by dengziming — APPROVED** [review]
LGTM.

**Review by satishd — APPROVED** [review]
Thanks @mimaison for the PR, LGTM.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+/*
+ * A log offset structure, including:
+ *  1. the message offset
+ *  2. the base message offset of the located segment
+ *  3. the physical position on the located segment
+ */
+public class LogOffsetMetadata {
+
+    //TODO remove once UnifiedLog has been moved to the storage module
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: Can we please use Jira tickets instead of TODOs? The latter tend to clutter the codebase and are often forgotten.

Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: Yes good call. I've added a reference to KAFKA-14484. Once merged I'll also comment in the JIRA.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        return this.segmentBaseOffset == that.segmentBaseOffset;
+    }
+
+    // compute the number of bytes between this offset to the given offset
+    // if they are on the same segment and this offset precedes the given offset
+    public int positionDiff(LogOffsetMetadata that) {
+        if (!onSameSegment(that))
+            throw new KafkaException(this + " cannot compare its segment position with " + that + " since they are not on the same segment");
+
```
Comment (fvaleri) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: Here you miss the `if(messageOffsetOnly)` check.

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 3]
```diff
@@ -444,13 +444,12 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
-          None
-        case newHighWatermark =>
-          Some(newHighWatermark)
-      }
+      val oldHighWatermark = highWatermarkMetadata.messageOffset()
+      val updatedHighWatermark = updateHighWatermark(new LogOffsetMetadata(hw))
+      if (updatedHighWatermark == oldHighWatermark)
+        None
+      else
+        Some(updatedHighWatermark)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: Is this change actually required?

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 4]
```diff
@@ -939,14 +938,14 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
-              appendInfo.firstOffset = Some(LogOffsetMetadata(duplicate.firstOffset))
+              appendInfo.firstOffset = Some(new LogOffsetMetadata(duplicate.firstOffset))
               appendInfo.lastOffset = duplicate.lastOffset
               appendInfo.logAppendTime = duplicate.timestamp
               appendInfo.logStartOffset = logStartOffset
             case None =>
               // Before appending update the first offset metadata to include segment information
               appendInfo.firstOffset = appendInfo.firstOffset.map { offsetMetadata =>
-                offsetMetadata.copy(segmentBaseOffset = segment.baseOffset, relativePositionInSegment = segment.size)
+                new LogOffsetMetadata(offsetMetadata.messageOffset(), segment.baseOffset, segment.size)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: Is it worth adding a method that keeps the same message offset, but updates the other two values? Something like `updateSegment`?

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala [hunk 5]
```diff
@@ -190,7 +190,7 @@ final class KafkaMetadataLog private (
   override def updateHighWatermark(offsetMetadata: LogOffsetMetadata): Unit = {
     offsetMetadata.metadata.asScala match {
       case Some(segmentPosition: SegmentPosition) => log.updateHighWatermark(
-        new kafka.server.LogOffsetMetadata(
+        new org.apache.kafka.server.log.internals.LogOffsetMetadata(
```
Comment (ijuma) [inline core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:193]: Why do we need to use the fully qualified name here?

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    //TODO KAFKA-14484 remove once UnifiedLog has been moved to the storage module
+    private static final long UNIFIED_LOG_UNKNOWN_OFFSET = -1L;
+
+    public static final LogOffsetMetadata UNKNOWN_OFFSET_METADATA = new LogOffsetMetadata(-1L, 0L, 0);
+
+    private static final int UNKNOWN_FILE_POSITION = -1;
+
+    private final long messageOffset;
+    private final long segmentBaseOffset;
+    private final int relativePositionInSegment;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: I think we should just make these public since they're final fields. We can then remove the accessors. For internal classes, it doesn't seem particularly useful to keep immutable fields private along with accessors that are public.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 2 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: Let's check if any test failed. If it did not, let's add coverage, please.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    // decide if the offset metadata only contains message offset info
+    public boolean messageOffsetOnly() {
+        return segmentBaseOffset == UNIFIED_LOG_UNKNOWN_OFFSET && relativePositionInSegment == UNKNOWN_FILE_POSITION;
+    }
+
+    @Override
+    public String toString() {
+        return "LogOffsetMetadata{" +
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: This differs from the previous `toString` implementation. Is that intentional?

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+import java.util.Objects;
+
+/*
+ * A log offset structure, including:
+ *  1. the message offset
+ *  2. the base message offset of the located segment
+ *  3. the physical position on the located segment
+ */
+public class LogOffsetMetadata {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: Let's make this final.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        LogOffsetMetadata that = (LogOffsetMetadata) o;
+        return messageOffset == that.messageOffset && segmentBaseOffset == that.segmentBaseOffset && relativePositionInSegment == that.relativePositionInSegment;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(messageOffset, segmentBaseOffset, relativePositionInSegment);
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: This implementation results in boxing all the values and then adding them to an array, etc. To avoid any regressions, we should go with the approach that doesn't do any of that.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Left a few comments.

**Comment by ijuma** [comment]
A couple more things: conflicts need to be resolved and is it the case that `LogOffsetMetadata.offsetDiff` is not used?

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 2 above)
Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: The `onSameSegment()` method already calls `messageOffsetOnly()` so this condition should be true if we reach this code.

File: core/src/main/scala/kafka/raft/KafkaMetadataLog.scala (hunk 5 above)
Comment (mimaison) [reply core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:193]: This class also uses `org.apache.kafka.raft.LogOffsetMetadata`. The argument of that method is actually of that type.

Comment (ijuma) [reply core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:193]: Ok, two classes with the same name are annoying in Java. I guess we'll have to see if we can consolidate later (as you said elsewhere).

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 2 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: That said the message won't be the same and it will perhaps be a bit confusing? That is, perhaps we should do the message offset check first.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 3 above)
Comment (mimaison) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: I was required previously as we can't have getters as an `case` option. Now that we're moving to public fields this change is not needed anymore, so I'll revert it.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 2 above)
Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: The `if(messageOffsetOnly)` check here was effectively unreachable code. If this was true, `onSameSegment()` would have thrown already (with the exact same message).

Since `onSameSegment()` was not marked private, I guess it made sense to have the check there too, "just in case". But this method is only called from `positionDiff()`, so I've marked it private now and thought we did not need to have the redundant/unreachable check.

Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java:72]: I understood that part. My point was that the original goal was to have a specific error message for this method even if that didn't work out due to the other method throwing first. If you do the message offset only check first, then the problem is solved.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 4 above)
Comment (mimaison) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: It's the only place where we do this. Apart if you have a strong opinion on this, I find using the constructor here fine.

**Comment by mimaison** [comment]
I've rebased and either addressed or replied to the comments.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    // check if this offset is on the same segment with the given offset
+    private boolean onSameSegment(LogOffsetMetadata that) {
+        return this.segmentBaseOffset == that.segmentBaseOffset;
+    }
+
+    // compute the number of bytes between this offset to the given offset
+    // if they are on the same segment and this offset precedes the given offset
+    public int positionDiff(LogOffsetMetadata that) {
+        if (messageOffsetOnly())
+            throw new KafkaException(this + " cannot compare its segment info with " + that + " since it only has message offset info");
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: "cannot compare its segment info" should be "cannot compare its segment position" (that was the difference between the message in `positionDiff` versus the one in `onSameSegment`.

**Review by ijuma — APPROVED** [review]
LGTM apart from the one comment above.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 10 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: I pushed this fix so that we can merge this PR.

**Comment by ijuma** [comment]
JDK 17 build passed and JDK 11 build had unrelated failures.

File: storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java (hunk 10 above)
Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java]: Sorry I was away yesterday. Thanks!
